libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -0,0 +1,271 @@
+        if (authResponse.isSuccess) {
+            val code = authResponse.asJSONObject().get(CODE).toString()
+            val authEndpoint = authResponse.asJSONObject().get(SFDC_COMMUNITY_URL).toString()
+            val tokenRequestBody = createRequestBody(
+                CODE to code,
+                GRANT_TYPE to AUTHORIZATION_CODE,
+                CLIENT_ID to clientId,
+                REDIRECT_URI to redirectUri,
+                CODE_VERIFIER to codeVerifier,
+            )
+            authResponse.consumeQuietly()
+
+            val tokenRequest = RestRequest(
+                RestRequest.RestMethod.POST,
+                RestRequest.RestEndpoint.LOGIN,
+                authEndpoint + OAUTH_TOKEN_PATH,  // Full path for unauthenticated request
+                tokenRequestBody,
+                 null, // additionalHttpHeaders
+            )
+
+            // Second REST Call - token request with code verifier
+            val tokenResponse = suspendedRestCall(tokenRequest) ?: return NativeLoginResult.UnknownError
+            return if (tokenResponse.isSuccess) {
+                // Returns NativeLoginResult.Success or UnknownError based on if the user is created.
+                suspendFinishAuthFlow(tokenResponse)
+            } else {
+                NativeLoginResult.UnknownError
+            }
+        } else {
+            SalesforceSDKLogger.e(TAG, "Native Login Authorization Error: $authResponse")
+            return NativeLoginResult.InvalidCredentials
+        }
+    }
+
+    override fun getFallbackWebAuthenticationIntent(): Intent {
+        val context = SalesforceSDKManager.getInstance().appContext
+        val intent = Intent(context, SalesforceSDKManager.getInstance().webviewLoginActivityClass)
+        intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
+        val options = SalesforceSDKManager.getInstance().loginOptions.asBundle()
+        options.putBoolean(BiometricAuthenticationManager.SHOW_BIOMETRIC, bioAuthLocked)
+        intent.putExtras(options)
+        Bundle().putParcelable(AccountManager.KEY_INTENT, intent)
+
+        return intent
+    }
+
+    @VisibleForTesting
+    internal fun isValidUsername(username: String): Boolean {
+        if (username.length > MAX_USERNAME_LENGTH) {
+            return false
+        }
+
+        return Regex(USERNAME_REGEX_PATTERN).matches(username)
+    }
+
+    @VisibleForTesting
+    internal fun isValidPassword(password: String): Boolean {
+        val containsNumber = password.contains("[0-9]".toRegex())
+        val containsLetter = password.contains("[A-Za-z]".toRegex())
+
+        return containsNumber && containsLetter && password.length >= MIN_PASSWORD_LENGTH
+                && password.toByteArray().size <= MAX_PASSWORD_LENGTH_BYTES
+    }
+
+    private suspend fun suspendFinishAuthFlow(tokenResponse: RestResponse): NativeLoginResult {
+        val appContext = SalesforceSDKManager.getInstance().appContext
```

Comment:
I believe we have to eliminate this Context reference in `SalesforceSDKManager` someday.  Should we have native login manager receive it another way now to make that easier?  Perhaps we could have the app provide it in the initialize native login method?  I might be thinking to far ahead on this one.